src/chainstate/stacks/index/storage.rs

```diff
@@ -0,0 +1,1213 @@
+/*
+ copyright: (c) 2013-2019 by Blockstack PBC, a public benefit corporation.
+
+ This file is part of Blockstack.
+
+ Blockstack is free software. You may redistribute or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License or
+ (at your option) any later version.
+
+ Blockstack is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY, including without the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with Blockstack. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use std::fmt;
+use std::error;
+use std::io;
+use std::io::{
+    Read,
+    Write,
+    Seek,
+    SeekFrom,
+    Cursor
+};
+
+use std::char::from_digit;
+use std::marker::PhantomData;
+use std::ops::{Deref, DerefMut};
+use hashbrown::HashMap;
+use std::collections::VecDeque;
+
+use std::fs;
+use std::path::{
+    Path,
+    PathBuf
+};
+
+use std::os;
+
+#[cfg(target_os = "unix")]
+use std::os::unix::io::AsRawFd;
+
+#[cfg(target_os = "unix")]
+use libc;
+
+use regex::Regex;
+
+use chainstate::burn::BlockHeaderHash;
+use chainstate::burn::BLOCK_HEADER_HASH_ENCODED_SIZE;
+
+use chainstate::stacks::index::{
+    TrieHash,
+    TRIEHASH_ENCODED_SIZE,
+    fast_extend_from_slice,
+};
+
+use chainstate::stacks::index::bits::{
+    get_node_byte_len,
+    write_nodetype_bytes,
+    read_hash_bytes,
+    read_node_hash_bytes,
+    read_nodetype,
+    get_node_hash,
+    trie_hash_from_bytes
+};
+
+use chainstate::stacks::index::node::{
+    is_backptr,
+    clear_backptr,
+    set_backptr,
+    TrieNodeType,
+    TrieNode4,
+    TrieNode16,
+    TrieNode48,
+    TrieNode256,
+    TrieLeaf,
+    TrieNodeID,
+    TriePtr,
+    TriePath,
+    TrieNode
+};
+
+use chainstate::stacks::index::fork_table::{
+    TrieForkPtr,
+    TrieForkTable
+};
+
+use chainstate::stacks::index::Error as Error;
+
+use util::log;
+
+pub fn ftell<F: Seek>(f: &mut F) -> Result<u64, Error> {
+    f.seek(SeekFrom::Current(0))
+        .map_err(Error::IOError)
+}
+
+pub fn fseek<F: Seek>(f: &mut F, off: u64) -> Result<u64, Error> {
+    f.seek(SeekFrom::Start(off))
+        .map_err(Error::IOError)
+}
+
+pub fn fseek_end<F: Seek>(f: &mut F) -> Result<u64, Error> {
+    f.seek(SeekFrom::End(0))
+        .map_err(Error::IOError)
+}
+
+pub fn read_all<R: Read>(f: &mut R, buf: &mut [u8]) -> Result<usize, Error> {
+    let mut cnt = 0;
+    while cnt < buf.len() {
+        let nr = f.read(&mut buf[cnt..])
+            .map_err(Error::IOError)?;
+
+        if nr == 0 {
+            break;
+        }
+
+        cnt += nr;
+    }
+    Ok(cnt)
+}
+
+pub fn write_all<W: Write>(f: &mut W, buf: &[u8]) -> Result<usize, Error> {
```

Comment:
The trait Write provides a `write_all` function -- we should just use that:

https://doc.rust-lang.org/std/io/trait.Write.html#method.write_all

Reply:
:+1: